### ISSUE TITLE
Expose SQS message's ReceiptHandle

### ISSUE THREAD

**Issue by simonness** [justeat/JustSaying#189]
I have a use case where it would be beneficial to call [ChangeMessageVisibility](http://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_ChangeMessageVisibility.html).

Seems like a big change to provide full support for this use case, an interim solution might be to expose an SQS message's ReceiptHandle on the Just Saying Message type so that the AWS SDK can at least be called directly?

**Comment by payman81** [comment]
That means all Message types will take dependency on some AWS type. How about having an envelope around the message that can contain metadata around the message such as ReceiptHandle, SQS message id etc... This way JustSaying Message remains intact.

```
    public class OrderPlaced : Message
    {
        public int OrderId { get; set; }
    }

    public class MessageEnvelope<T> where T : Message
    {
        public T Message { get; set; }
        public string SqsReceiptHandle { get; set; }
        public Guid SqsMessageId { get; set; }

    }
```

**Comment by payman81** [comment]
Then handlers' interface would change slightly to:
```
    public class MessageHandler : IHandler<OrderPlaced>
    {
        public bool Handle(MessageEnvelope<OrderPlaced> envelope)
        {
            var message = envelope.Message;

            throw new NotImplementedException();
        }
    }
```

**Comment by petemounce** [comment]
Can we have our hash of meta data via this?

--

**Comment by payman81** [comment]
Yes @petemounce.

**Comment by dariusdamalakas** [comment]
Interesting problem. I like the envelope suggestion - we can move this even further and make all message class POCO, and even potentially remove JustSaying.Models altogether (need to validate this).

Another approach could be to introduce IMessage, again with headers dictionary to contain any extra data we want to expose, such as receip handles, etc.

**Comment by AnthonySteele** [comment]
I like the Envelope idea, but why does JustSaying need to do any additional parsing of the SQS message? How about:

````csharp
    public class MessageEnvelope<T>
    {
        public Amazon.SQS.Model.Message Metadata { get; set; }
        public T Message { get; set; }
    }

 
    // if you want to work with the message as before, 
    // you just need to  extract "env.Message" , so an overhead of 1 line of code
    public interface IEnvelopeHandler<T>
    {
        Task<bool> Handle(MessageEnvelope<T> env);
    }
    
`````

**Comment by ssemyonov** [comment]
@payman81 somehting like #239 could solve the issue without significantly changing current interface and adding dependency on AWS model.
Any reasons not adding ```public string ReceiptHandle { get; set; }``` to ```Message``` class?

**Comment by payman81** [comment]
@ssemyonov The only reason I can think of is mixing message payload and metadata about the message. However, Message type already contains metadata e.g. Tenant so I don't think this makes it a lot worse. 
@AnthonySteele What do you think?

**Comment by simonness** [comment]
Fixed by #240 and #242.  Thanks @ssemyonov and @payman81 :)
